File: java/438.find-all-anagrams-in-a-string.java
/*
 * @lc app=leetcode id=438 lang=java
 *
 * [438] Find All Anagrams in a String
 */
class Solution {

    int[] ans = new int[26];

    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> res = new LinkedList<>();
        if (s == null || p == null || s.length() < p.length()) return res;
        int[] freq = new int[26];
        for (int t = 0; t < p.length(); ++t) {
            ans[p.charAt(t)-'a']++;
            freq[s.charAt(t)-'a']++;
        }
        for (int i = 0; ; ++i) {
            if (isAnagram(freq)) res.add(i);
            if (i == s.length() - p.length()) break;
            freq[s.charAt(i)-'a']--;
            freq[s.charAt(i+p.length())-'a']++;
        }
        return res;
    }

    private boolean isAnagram(int[] freq) {
        for (int i = 0; i < 26; ++i) {
            if (freq[i] != ans[i]) return false;
        }
        return true;
    }
}

File: go/438.find-all-anagrams-in-a-string.go
package main

/*
 * @lc app=leetcode id=438 lang=golang
 *
 * [438] Find All Anagrams in a String
 */

// @lc code=start
func findAnagrams(s string, p string) []int {
	res := make([]int, 0)
	if len(s) < len(p) {
		return res
	}
	const a = int('a')
	var t [26]int
	for i := 0; i < len(p); i++ {
		t[int(p[i])-a] += 1
		t[int(s[i])-a] -= 1
	}
	if isAna(&t) {
		res = append(res, 0)
	}
	for i := len(p); i < len(s); i++ {
		t[int(s[i])-a] -= 1
		t[int(s[i-len(p)])-a] += 1
		if isAna(&t) {
			res = append(res, i-len(p)+1)
		}
	}
	return res
}

func isAna(t *[26]int) bool {
	for i := 0; i < 26; i++ {
		if (*t)[i] != 0 {
			return false
		}
	}
	return true
}

// @lc code=end
